File: Video Player/VPSearchResultController.m
//
//  VPSearchResultController.m
//  
//
//  Created by 朱 文杰 on 15/6/19.
//
//

#import "AppDelegate.h"
#import "VPSearchResultController.h"
#import <BlocksKit+UIKit.h>
#import <TOWebViewController/TOWebViewController.h>

@implementation VPSearchResultController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = [[NSString alloc] initWithFormat:@"%@: %@ (%lu)", NSLocalizedString(@"Search", @"Search"), self.keyword, (unsigned long)[self.torrents count]];
    NSRegularExpression *regex = [[NSRegularExpression alloc] initWithPattern:@"^[A-Za-z]{2,6}-\\d{2,4}$" options:NSRegularExpressionCaseInsensitive error:nil];
    if ([regex matchesInString:self.keyword options:NSMatchingAnchored range:NSMakeRange(0, [self.keyword length])] > 0) {
        __weak typeof(self) weakself = self;
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] bk_initWithImage:[UIImage imageNamed:@"wiki"] style:UIBarButtonItemStylePlain handler:^(id sender) {
            TOWebViewController *webViewController = [[TOWebViewController alloc] initWithURLString:[NSString stringWithFormat:@"http://www.javlibrary.com/cn/vl_searchbyid.php?keyword=%@", [[self.keyword stringByReplacingOccurrencesOfString:@" " withString:@""] stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]]];
            webViewController.showUrlWhileLoading = NO;
            webViewController.hidesBottomBarWhenPushed = YES;
            [weakself.navigationController pushViewController:webViewController animated:YES];
        }];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (self.navigationController.navigationBarHidden) {
        [self.navigationController setNavigationBarHidden:NO animated:YES];
    }
    if ([UIApplication sharedApplication].statusBarHidden) {
        [[UIApplication sharedApplication] setStatusBarHidden:NO withAnimation:UIStatusBarAnimationSlide];
    }
    self.navigationController.navigationBar.tintColor = nil;
    [self.navigationController.navigationBar setBarStyle:UIBarStyleDefault];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation {
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone)
        return toInterfaceOrientation != UIInterfaceOrientationMaskPortraitUpsideDown;
    else
        return YES;
}

#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [self.torrents count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"FileListTableViewCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
    }
    NSDictionary *torrent = self.torrents[indexPath.row];
    cell.textLabel.text = torrent[@"name"];
    
    id size = torrent[@"size"];
    NSString *sizeString = NSLocalizedString(@"Unknown size", @"Unknown size");
    if (![size isKindOfClass:[NSNull class]]) {
        long long sizeValue = [size longLongValue];
        sizeString = [self stringForSize:sizeValue];
    }
    
    NSString *dateString = [self formattedDate:[torrent[@"upload_date"] integerValue]];
    cell.detailTextLabel.text = [[NSString alloc] initWithFormat:@"%@, %@%@, %@", sizeString, torrent[@"seeders"], NSLocalizedString(@"seeders", @"seeders"), dateString];
    cell.accessoryType = UITableViewCellAccessoryDetailDisclosureButton;
    return cell;
}

#pragma mark - Table view delegate
- (void)tableView:(UITableView *)tableView accessoryButtonTappedForRowWithIndexPath:(NSIndexPath *)indexPath {
    NSDictionary *torrent = self.torrents[indexPath.row];
    [self addTorrentToTransmission:torrent];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSDictionary *torrent = self.torrents[indexPath.row];
    UIAlertView *alert = [[UIAlertView alloc] bk_initWithTitle:NSLocalizedString(@"Info", @"Info") message:[self torrentDescription:torrent]];
    __weak typeof(self) weakself = self;
    [alert bk_addButtonWithTitle:NSLocalizedString(@"Download", @"Download") handler:^{
        [weakself addTorrentToTransmission:torrent];
    }];
    [alert bk_setCancelButtonWithTitle:NSLocalizedString(@"OK", @"OK") handler:nil];
    [alert show];
}

- (NSString *)formattedDate:(NSInteger)timeStamp {
    NSDate *date = [NSDate dateWithTimeIntervalSince1970:timeStamp];
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    NSLocale *locale = [NSLocale currentLocale];
    [formatter setLocale:locale];
    [formatter setDateStyle:NSDateFormatterShortStyle];
    [formatter setTimeStyle:NSDateFormatterShortStyle];
    return [formatter stringFromDate:date];
}

- (NSString *)torrentDescription:(NSDictionary *)dict {
    return [[NSString alloc] initWithFormat:@"%@, \n"
                                             "%@, \n"
                                             "%@, \n"
                                             "%@, \n"
                                             "%@"
                                             " seeders",
                                            dict[@"name"],
                                            [self stringForSize:[dict[@"size"] longLongValue] ],
                                            dict[@"magnet"],
                                            [self formattedDate:[dict[@"upload_date"] integerValue]],
                                            dict[@"seeders"],
                                            NSLocalizedString(@"seeders", @"seeders")];
}

- (NSString *)stringForSize:(long long)size {
    NSString *sizeString = @"";
    double mbValue = size / (1024.0 * 1024.0);
    if (mbValue > 1024.0) {
        sizeString = [[NSString alloc] initWithFormat:@"%.2f GB", mbValue / 1024.0];
    }
    else {
        sizeString = [[NSString alloc] initWithFormat:@"%.1f MB", mbValue];
    }
    return sizeString;
}

- (void)addTorrentToTransmission:(NSDictionary *)torrent {
    __weak typeof(self) weakself = self;
    [[AppDelegate shared] parseSessionAndAddTask:torrent[@"magnet"] completionHandler:^{
        [[AppDelegate shared] showHudWithMessage:NSLocalizedString(@"Task added.", @"Task added.") inView:weakself.navigationController.view];
    } errorHandler:^{
        [[AppDelegate shared]  showHudWithMessage:NSLocalizedString(@"Unknow error.", @"Unknow error.") inView:weakself.navigationController.view];
    }];
}

@end

File: Video Player/VPSearchResultController.swift
//
//  VPSearchResultController.swift
//  Video Player
//
//  Created by 朱文杰 on 15/11/3.
//  Copyright © 2015年 Home. All rights reserved.
//

import UIKit
import TOWebViewController

class VPSearchResultController: UITableViewController {
    let CellIdentifier = "FileListTableViewCell"
    var torrents: [[String:AnyObject]] = []
    var keyword: String = ""
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = String(format: NSLocalizedString("%@: %@ (%lu)", comment: "%@: %@ (%lu)"), arguments: [NSLocalizedString("Search", comment:"Search"), keyword, torrents.count])

        if keyword.matches("^[A-Za-z]{2,6}-\\d{2,6}$", regularExpressionOptions: [.CaseInsensitive], matchingOptions:[.Anchored]) {
            navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(named: "wiki"), style: .Plain, target: self, action: "showWiki")
        }

        // Theme
        navigationController?.navigationBar.barTintColor = Helper.defaultHelper.mainThemeColor()
        navigationController?.navigationBar.tintColor = UIColor.whiteColor()
        navigationController?.navigationBar.titleTextAttributes = [NSForegroundColorAttributeName : UIColor.whiteColor()]
        tableView.tintColor = Helper.defaultHelper.mainThemeColor()

        // Revert back to old UITableView behavior
        if #available(iOS 9.0, *) {
            tableView.cellLayoutMarginsFollowReadableWidth = false
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)

        if UIApplication.sharedApplication().statusBarHidden {
            UIApplication.sharedApplication().setStatusBarHidden(false, withAnimation: .Slide)
        }
        
        navigationController?.setToolbarHidden(true, animated: true)
        navigationController?.navigationBar.barStyle = .Default
    }

    override func preferredStatusBarStyle() -> UIStatusBarStyle {
        return .LightContent
    }

    // MARK: - Table view delegate and data source
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return torrents.count
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell: UITableViewCell! = tableView.dequeueReusableCellWithIdentifier(CellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .Subtitle, reuseIdentifier: CellIdentifier)
        }

        // Configure the cell...
        let torrent = torrents[indexPath.row]
        cell.textLabel?.text = torrent["name"] as? String
        cell.accessoryType = .DetailDisclosureButton;

        let size = convertSizeToString(torrent["size"])
        let dateString = formattedDateString(torrent["upload_date"] as? Int)
        let seeders = torrent["seeders"] as? Int
        cell.detailTextLabel?.text = "\(size), \(seeders == nil ? 0 : seeders!)\(NSLocalizedString("seeders", comment:"seeders")), \(dateString)"

        return cell
    }

    override func tableView(tableView: UITableView, accessoryButtonTappedForRowWithIndexPath indexPath: NSIndexPath) {
        let torrent = torrents[indexPath.row]
        self.addTorrentToTransmission(torrent)
    }

    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        let torrent = torrents[indexPath.row]
        let alertController = UIAlertController(title: NSLocalizedString("Info", comment: "Info"), message: describe(torrent), preferredStyle: .Alert)
        let addTorrentAction = UIAlertAction(title: NSLocalizedString("Download", comment: "Download") , style: .Default) { [unowned self] _ in
            self.addTorrentToTransmission(torrent)
        }
        alertController.addAction(addTorrentAction)
        let cancelAction = UIAlertAction(title: NSLocalizedString("OK", comment: "OK"), style: .Cancel, handler: nil)
        alertController.addAction(cancelAction)
        presentViewController(alertController, animated: true, completion: nil)
    }

    //MARK: - Action
    func showWiki() {
        let webViewController = TOWebViewController(URLString: "http://www.javlib3.com/cn/vl_searchbyid.php?keyword=\(keyword)")
        webViewController.showUrlWhileLoading = false
        webViewController.hidesBottomBarWhenPushed = true
        webViewController.loadingBarTintColor = Helper.defaultHelper.mainThemeColor()
        webViewController.buttonTintColor = Helper.defaultHelper.mainThemeColor()
        navigationController?.pushViewController(webViewController, animated: true)
    }

    //MARK: - Helper
    func describe(torrent: NSDictionary) -> String {
        let name = torrent["name"] as! String
        let size = convertSizeToString(torrent["size"])
        let magnet = torrent["magnet"] as! String
        let date = formattedDateString((torrent["upload_date"] as? Int))
        let seeders = torrent["seeders"] as? Int
        return "\(name), \n\(size), \n\(magnet), \n\(date), \n\(seeders == nil ? 0 : seeders!) " + NSLocalizedString("seeders", comment:"")
    }

    func formattedDateString(timeStamp:Int?) -> String {
        guard let ts = timeStamp else { return NSLocalizedString("Unknown date", comment: "Unknown date") }
        let date = NSDate(timeIntervalSince1970: Double(ts))
        let formatter = NSDateFormatter()
        formatter.locale = NSLocale.currentLocale()
        formatter.dateStyle = .ShortStyle
        formatter.timeStyle = .ShortStyle
        return formatter.stringFromDate(date)
    }

    func convertSizeToString(size:AnyObject?) -> String {
        let number = size as! NSNumber
        return number.longLongValue.fileSizeString
    }

    func addTorrentToTransmission(torrent: [String:AnyObject]) {
        Helper.defaultHelper.parseSessionAndAddTask(torrent["magnet"] as! String, completionHandler: {
            Helper.defaultHelper.showHudWithMessage(NSLocalizedString("Task added.", comment: "Task added."))
        }, errorHandler: {
            Helper.defaultHelper.showHudWithMessage(NSLocalizedString("Unknow error.", comment: "Unknow error."))
        })
    }
}
